### PR TITLE
3016 list add pinned items

### LINKED ISSUE #3016: [List] add pinned items
### What package is this for?

- [x] React
- [ ] Angular

### Summary
As part of #2603 we need to be able to lock an item so that it can't be reordered in the list. 

**Additional context** 

image

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release?

_NB: The core contributors will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### Available extra resources

What resources do you have to assist this effort?

_Carbon add-ons for Watson IoT is a collaborative system. We encourage teams to
build components and submit them for integration._

### PR DESCRIPTION
**Pull request by bjornalm**
Closes #3016 

**Summary**

Added feature to lock rows and prevent them from being reordered

**Change List (commits, features, bugs, etc)**
- Added `lockedIds` prop to List and HierarchyList
- Added List story showing `lockedIds`
- Updated Hierarchy story to show knob for `lockedIds`

**Acceptance Test (how to verify the PR)**

1. Go to https://deploy-preview-3017--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-nested-reorder
2. Verify that Jeff McNeil can't be directly reordered and that there is a lock icon on the row
3. Turn on multi-nesting and make sure it looks and works ok
4. Turn on hasVirtualList and make sure it looks and works ok

1. Go to https://deploy-preview-3017--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-list--with-reorder-and-locked-rows
2. Verify that the two top rows can't be directly reordered and that there is a lock icon on the rows
3. Verify that it is not possible to drag and drop another row above the locked row
4. Turn on hasVirtualList and make sure it looks and works ok

**Regression Test (how to make sure this PR doesn't break old functionality)**

All normal List and Hierarchy List stories work as before

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx [hunk 1]
```diff
@@ -799,6 +802,34 @@ describe('HierarchyList', () => {
...
+    expect(
+      within(screen.getByTestId('list'))
+        .getByText('Item 1')
+        .closest(`.${iotPrefix}--list-item-parent > *`)
+    ).not.toHaveAttribute('draggable');
+
+    expect(
+      within(screen.getByTestId('list')).getByText('Item 1').closest(`.${iotPrefix}--list-item`)
+        .firstChild
+    ).toHaveClass('iot--list-item--lock');
```
Comment (kevinsperrine) [inline packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:824]:
```suggestion
    ).toHaveClass(`${iotPrefix}--list-item--lock`);
```

File: packages/react/src/components/List/List.test.jsx [hunk 2]
```diff
@@ -363,6 +364,21 @@ describe('List', () => {
...
   });
 
+  it('should show lock icons and prevent rows from being dragged for ids in lockedIds', () => {
+    render(<List items={getListItems(2)} editingStyle={EditingStyle.Single} lockedIds={['1']} />);
+    expect(
+      screen.getByText('Item 1').closest(`.${iotPrefix}--list-item-parent > *`)
+    ).not.toHaveAttribute('draggable');
+
+    expect(screen.getByText('Item 1').closest(`.${iotPrefix}--list-item`).firstChild).toHaveClass(
+      'iot--list-item--lock'
```
Comment (kevinsperrine) [inline packages/react/src/components/List/List.test.jsx:374]:
```suggestion
      `${iotPrefix}--list-item--lock`
```

File: packages/react/src/components/List/List.test.jsx [hunk 3]
```diff
@@ -760,5 +776,27 @@ describe('List', () => {
...
+          editingStyle={EditingStyle.Single}
+          lockedIds={['1']}
+        />
+      );
+      expect(
+        screen.getByText('Item 1').closest(`.${iotPrefix}--list-item-parent > *`)
+      ).not.toHaveAttribute('draggable');
+
+      expect(screen.getByText('Item 1').closest(`.${iotPrefix}--list-item`).firstChild).toHaveClass(
+        'iot--list-item--lock'
```
Comment (kevinsperrine) [inline packages/react/src/components/List/List.test.jsx:794]:
```suggestion
        `${iotPrefix}--list-item--lock`
```

**Review by kevinsperrine — APPROVED** [review]
Nice work! Looks great to me. Just a couple `iotPrefix` replacements, and a couple questions:

Should we still be able to move a locked row in multiple-nesting using the checkbox in the [HierarchyList with nested reorder story](https://deploy-preview-3017--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-nested-reorder)?

Do we need to test a scenario with a locked rows that items cannot be dragged above in the HierarchyList (ie. scenario 2 from the acceptable instructions, but with a HierarchyList instead)?

**Comment by bjornalm** [comment]
> Nice work! Looks great to me. Just a couple `iotPrefix` replacements, and a couple questions:
> 
> Should we still be able to move a locked row in multiple-nesting using the checkbox in the [HierarchyList with nested reorder story](https://deploy-preview-3017--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-nested-reorder)?

I don't know, probably not. It is not an issue for #2603 since it doesn't use the multiple-nesting but I will disable that checkbox if the row is locked since that makes more sense. Good call.

> Do we need to test a scenario with a locked rows that items cannot be dragged above in the HierarchyList (ie. scenario 2 from the acceptable instructions, but with a HierarchyList instead)?

That is a valid question. There is no clear spec for this as a general List/HierarchyList functionality but you can test the scenario by using the knob `lockedIds` in the story "1-watson-iot-hierarchylist--with-nested-reorder". In order to prevent anything from being dragged above a topmost locked child item the parent needs to be locked as well. That is okay for my purpose with #2603.

**Comment by davidicus** [comment]
It could be a storybook oddity but... If you turn on multi-select and you have the initial selected item still selected and then choose one more item. Move them both and only the second item is still selected. 
| Initial state | choose a second | once moved |
|-----|---------|------|
| image | image | image

**Comment by bjornalm** [comment]
> It could be a storybook oddity but... If you turn on multi-select and you have the initial selected item still selected and then choose one more item. Move them both and only the second item is still selected.
> 
I think this is a bug in the component, but it is unrelated to this PR and also happens in next. I created a new issue for it here https://github.com/carbon-design-system/carbon-addons-iot-react/issues/3042

**Comment by bjornalm** [comment]
> The two locked rows do not show as draggable but I am unable to drag and drop any rows on the second link

I think you can drag all the other items (and that should be made visible by the blue line that indicates the drop zones) but nothing will happen once you drop since the List is stateless. The HierarchyList is stateful and that story also shows what happens after the actual drop. Let me know if I misunderstood you and you actually can't actually drag and drop any rows at all.

**Comment by kevinsperrine** [comment]
> > The two locked rows do not show as draggable but I am unable to drag and drop any rows on the second link
> 
> I think you can drag all the other items (and that should be made visible by the blue line that indicates the drop zones) but nothing will happen once you drop since the List is stateless. The HierarchyList is stateful and that story also shows what happens after the actual drop. Let me know if I misunderstood you and you actually can't actually drag and drop any rows at all.

We could add an action to the onItemMoved callback to make it clear something is happening.

**Comment by bjornalm** [comment]
> > > The two locked rows do not show as draggable but I am unable to drag and drop any rows on the second link
> > 
> > 
> > I think you can drag all the other items (and that should be made visible by the blue line that indicates the drop zones) but nothing will happen once you drop since the List is stateless. The HierarchyList is stateful and that story also shows what happens after the actual drop. Let me know if I misunderstood you and you actually can't actually drag and drop any rows at all.
> 
> We could add an action to the onItemMoved callback to make it clear something is happening.

Yes, good idea. I'll add that.

**Review by davidicus — APPROVED** [review]

**Pull request merged**